Domains/Jobs/WorkerConfiguration.php

```diff
@@ -0,0 +1,338 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * @file Domains/Jobs/Job.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class Job
+ * @ingroup domains
+ *
+ * @brief Laravel Eloquent model for Jobs table
+ */
+
+namespace PKP\Domains\Jobs;
+
+class WorkerConfiguration
+{
+    /**
+     * The name of the worker.
+     *
+     * @var string
+     */
+    protected $name = 'default';
+
+    /**
+     * Calculate the number of seconds to wait before retrying the job.
+     *
+     * @var int
+     */
+    protected $backoff = 0;
+
+    /**
+     * Job's worker allowed memory configuration, as the maximum amount of RAM (in megabytes) the worker may consume.
+     *
+     * @var int
+     */
+    protected $memory = 128;
+
+    /**
+     * Job's worker timeout configuration, as the maximum number of seconds a child worker may run.
+     *
+     * @var int
+     */
+    protected $timeout = 30;
+
+    /**
+     * Job's worker sleep configuration, as the number of seconds to wait/sleep in between polling the queue.
+     *
+     * @var int
+     */
+    protected $sleep = 3;
+
+    /**
+     * Number of times to attempt a job before logging it failed
+     *
+     * @var int
+     */
+    protected $tries = 3;
+
+    /**
+     * Job's worker force configuration, indicates if the worker should run even in Laravel's maintenance mode.
+     *
+     * @var bool
+     */
+    protected $force = false;
+
+    /**
+     * Job's worker stopWhenEmpty configuration, indicates if the worker should stop when queue is empty.
+     *
+     * @var bool
+     */
+    protected $stopWhenEmpty = false;
+
+    /**
+     * The number of jobs to process before stopping
+     *
+     * @var int
+     */
+    protected $maxJobs = 0;
+
+
+    /**
+     * The maximum number of seconds the worker should run
+     *
+     * @var int
+     */
+    protected $maxTime = 0;
+
+
+    /**
+     * Number of seconds to rest between jobs
+     *
+     * @var int
+     */
+    protected $rest = 0;
+
+    /**
+     * Set the configurations
+     *
+     * @param array $options
+     * @return self
+     */
+    public static function withOptions(array $options = []) : self
+    {
+        $self = new static;
+
+        collect($options)
+            ->each(fn($value, $option) => method_exists($self, 'set' . ucfirst($option)) 
+                ? $self->{'set' . ucfirst($option)}($value) 
+                : null
+            );
+        
+        return $self;
+    }
+
+    /**
+     * Get the worker options
+     *
+     * @return array
+     */
+    public function getWorkerOptions() : array 
+    {
+        return [
```

Comment:
I wonder if this could be rewritten to use named arguments (and an associative array) rather than a long positional list.

Reply:
@asmecher  returning back an `associative` make sense and updated accordingly. But I am bit confused about the part `rewritten to use named arguments`.  Can you provide bit more details about what kind of changes you are thinking ?

Reply:
When this gets passed into the `WorkerOptions` constructor via the splat (`...`) operator, the array gets unpacked into a long list of parameters. We're trying to avoid long positional parameter lists in function calls, especially where the positions of them are arbitrary. Making this an associative array, and using named arguments, makes it less likely that some of the parameters could e.g. get put in the list in the wrong order since the function call will depend on string keys rather than array positions.

Unfortunately, now that I search it, applying `...` to keyed arrays is only available [starting with PHP8.1](https://php.watch/versions/8.1/spread-operator-string-array-keys) -- so on PHP8.0 I think [this line](https://github.com/pkp/pkp-lib/pull/8133/files#diff-d8bccac695c8cacb01f4be531434b57599b8747080dd26b56e8f1cb6d5568274R68) will raise an error. But I think you could work around that using https://www.php.net/manual/en/function.call-user-func-array.php.

Reply:
> applying `...` to keyed arrays is only available [starting with PHP8.1](https://php.watch/versions/8.1/spread-operator-string-array-keys)

weirdly it is working on my set up which use `PHP 8.0.19` so I thought it's added to PHP8.0+. 

> But I think you could work around that using https://www.php.net/manual/en/function.call-user-func-array.php.

I don't think using the `call_user_func_array` is the right one here for us as it's take first param as a `callable` but we need to pass those params to the `\Illuminate\Queue\WorkerOptions::__construct` which has no other method to set those after the instance generated . I thought about extending the `\Illuminate\Queue\WorkerOptions` to have a static constructor so that we can passed named agrs but that seems like a bit too much workaround. Also the `PKP\Domains\Jobs\WorkerConfiguration::getWorkerOptions` set the params ordering to make it compatible with `\Illuminate\Queue\WorkerOptions::__construct` and that is why we have that class so we should not directly pass params to `\Illuminate\Queue\WorkerOptions` but build up those through the `PKP\Domains\Jobs\WorkerConfiguration`.  For now to make it compatible with PHP8.0+, I have extracted the values through `array_values`. 

Or may be I am missing something ?

Reply:
Hmm, you're right -- unpacking associative arrays as named arguments does appear to work with PHP8.0; I've tested with 8.0.2, which is pretty old. I think we're OK to continue with what you've proposed :+1: